### DISCUSSION TITLE
Presentations

### DISCUSSION THREAD

**Discussion by wmannis** [rabbibotton/clog#20] [Ideas]
Is there any plan to include CLIM and Genera-style presentations?

If you're unfamiliar, anything that goes to the screen can have a presentation type associated with it. Things like mouse actions can be bound to presentation types, rather than screen-element by screen-element. It also means that if you ask for input of a certain presentation type, they'll highlight as being available for a click.

**Comment by rabbibotton** [comment]
My initial goal (for the most part met at this point) was to create a low level binding to html and JavaScript. My next step beyond another demo or two and some filling in some blanks on the binding will be in the way of visual tools. I have considered adding higher level approaches but have not settled on anything concrete yet (at least for me to do, hopefully others will take and consider extending CLOG as well). 

Do you have any links to some sample code or tutorials for either of the approaches you mentioned? I’d love to become familiar with them.

**Reply by wmannis** [reply]
[McCLIM](https://github.com/McCLIM/McCLIM) is a recent implementation of CLIM. The Listener example application is a clear demonstration of the principles. If you run it, and list a directory, you'll notice that all the files listed can be interacted with.

I've not dug too deeply into CLOG yet, but I have a niche hobby app that I started to move to McCLIM but which I think would be more useful as a web app. I don't use presentations heavily, but it'd be useful to have some similar features eventually.

**Reply by rabbibotton** [reply]
My main drivers are Macs, so only played with McClim for a few minutes one, but I am going to spend some time on the docs and code as sounds interesting and similar to a few projects I did 20 years ago. I usually don't find data driven UI approaches work out well, they get you up and running quickly but you crash and burn once you want to play with look and feel but willing to see McClim and see if can do one better :)

**Reply by wmannis** [reply]
Not everything that appears on the screen has to be wrapped in a presentation type, fortunately. [This video](https://www.youtube.com/watch?v=kfBmRsPRdGg) has some examples, including a lot of interaction with the Listener.

A low-key implementation of a similar idea is Tk's event tags in [text widgets](https://tkdocs.com/tutorial/text.html). The additional benefit of presentations is that they are types and can have a class hierarchy, and when you're asking the user for data, if you're asking for a presentation type, anything that matches that currently on the screen will highlight, thanks to the `CLIM:ACCEPT` method.

I'm not any sort of web dev expert, but I'd guess this could be handled with particular elements getting class annotations that event handlers could interrogate and dispatch from there. 

I realize you have your own priorities with CLOG, but I wanted to mention presentations as a useful GUI abstraction in Lisp's heritage (IMHO, of course).

**Reply by rabbibotton** [reply]
<<I realize you have your own priorities with CLOG, but I wanted to mention presentations as a useful GUI abstraction in Lisp's heritage (IMHO, of course).>>

The suggestions that deal with Lisp idioms are the most important to me! I have certain needs for CLOG as I need a functional cross platform GUI (McClim is not realistic on any platform I need) but CLOG is already there. Most everything now is higher order APIs. So anything that can make things lispier I am more then all ears :)

**Comment by rabbibotton** [comment]
Thank you for the video and direction. It is certainly very doable on top of CLOG, but having trouble thinking of how I would use the feature in say a business app?

**Reply by wmannis** [reply]
For very forms-centric applications, it probably wouldn't be very useful.

For more dashboard-like applications you could have actions in one graph area, say, require clicks in another graph area to compare data. The presentation class hierarchy could limit which areas are legal for the requested comparison. 

Any place where several dependent steps have to be taken to request a complex action could potentially have some of the dependencies managed by presentation classes (at least if some representation of the dependency is on the screen already).

I confess my use isn't business directed, so I've not though deeply about this specific area.

**Reply by rabbibotton** [reply]
I am going to let is stew in me for awhile.

I see something like -
creating a data integrity layer on a database
presentation layer :present on the same database
presentation layer :accept on to a form
something that models process (including approvals) would fit in there as well and cover your use-case.
(sorry that is just some raw thought for myself later but maybe makes sense to you)

**Reply by ramenbytes** [reply]
Page 1 & 2 from [this manual](https://common-lisp.net/project/mcclim/static/documents/mcclim.pdf) explain a little about how presentations are used. If you ask around on reddit and irc, I'm sure there are some Lispers with experience using presentation-based interfaces that would be willing to explain further. Unfortunately, I'm not knowledgeable enough to do so myself.

A related idea: After Daniel finishes documenting how to write McCLIM backends, maybe it will be possible to write one in CLOG. That could be interesting.

**Reply by rabbibotton** [reply]
Thanks!

I think worth a shot if possible to use clog for a backend! I am sure I will hear about it but let me know when that doc is ready.

**Reply by wmannis** [reply]
> A related idea: After Daniel finishes documenting how to write McCLIM backends, maybe it will be possible to write one in CLOG. That could be interesting.

When I saw the comparison of the browser to X11 I had the very same idea. I have one teeny bit of code that went into McCLIM, and that took me a while to figure out. I'm sadly not the person to do that sort of backend work, though it'd be very cool to see.

**Comment by rabbibotton** [comment]
I am working through plans now to incorporate some of the ideas of presentations.

I don't know how close in syntax to CLIM it will be as I want something simple like (create-presentation clog-obj any-class-obj) and (create-presentation-item clog-obj any-class-obj any-slot) to instantly associate any class with visual elements. 

The idea would be to quickly turn a class in to a form/visual rep, use CLOS before and after to handle initializations and validations, everything remains pure data/model.

I'd like that when I change a slot in code the change is on screen, when on screen is already in the object in memory (post validation).

All very doable with CL and the CLOG approach. This will also be the means to hook in to database and data serialization.

I like the idea of methods also instantly becoming buttons and other actionable controls.

**Reply by wmannis** [reply]
Marvelous! I was just wondering recently what would be involved in attaching presentations to certain elements in the CLOG Builder.

**Reply by rabbibotton** [reply]
Take a look at tutorial 29 and clog-presentations.lisp there is more work to do but that offers the very cool transparent link between lisp objects and the gui.

**Reply by rabbibotton** [reply]
btw if you have specific things examples of functionality etc now would be the time to shout :)

**Comment by ramenbytes** [comment]
Woohoo! I’ll try to take a look soon. Neat stuff.

**Comment by MCSH** [comment]
Thank you for that update, I looked around and it was neat, but here's a use case I have that might be of interest (I can open a new discussion if it isn't the right place).

I want to have a "command" like "process_x_and_y" that as input takes two objects, one of type X and one of type Y. I like to be able to activate the command (through any means, maybe a click on a button) which brings two input fields, user can either input elements of X and Y by typing them, or can click on any element of type X on the screen and any element of type Y on the screen. It can certainly be done on top of clog at the moment, but a built-in method would make it much simpler.

This code in clim would be something like...

```lisp
(define-presentation-type X ())
(define-presentation-type Y ())

(define-foo-command (comm-process_x_and_y :name t :menu t)
  ((x 'X) (y 'Y)) ;; input x of type X and y of type Y
  (do-stuff x y))
```

**Reply by rabbibotton** [reply]
I can't really see much use in modern GUIs outside of menus that change based on selected content, but that same idea could be extended. I will be looking at extending the menu system in clog-gui soon and will try and come up with a generalized system.
